Add TP-RMSE normalized error metric

jmartinpizarro opened the pull request:
<!--
Thank you 🙏 for your contribution to [Ultralytics](https://www.ultralytics.com/) 🚀! Your effort in enhancing our repositories is greatly appreciated. To streamline the process and assist us in integrating your Pull Request (PR) effectively, please follow these steps:

1. Check for Existing Contributions: Before submitting, kindly explore existing PRs to ensure your contribution is unique and complementary.
2. Link Related Issues: If your PR addresses an open issue, please link it in your submission. This helps us better understand the context and impact of your contribution.
3. Elaborate Your Changes: Clearly articulate the purpose of your PR. Whether it's a bug fix or a new feature, a detailed description aids in a smoother integration process.
4. Ultralytics Contributor License Agreement (CLA): To uphold the quality and integrity of our project, we require all contributors to sign the CLA. Please confirm your agreement by commenting below:

    I have read the CLA Document and I sign the CLA

For more detailed guidance and best practices on contributing, refer to our ✅ [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Your adherence to these guidelines ensures a faster and more effective review process.
-->

This PR is related to the issue https://github.com/ultralytics/ultralytics/issues/22955 and the sub-issue related with it.

Introduce an RMSE, TP-based normalised metric in order to measure the evolution of the offset of the TP boxes prediction with respect to the GT labels for the YOLO Predict model.

Useful in several cases described in the issue commented before.

My only main concern is about the calculation of the offset, in the file `ultralytics/models/yolo/detect/val.py`. I am currently computiing the iou and then assuming an mAP@50 for obtaining the TP between the GT and the prediction. 

I did this because of not wanting to modify the _process_batch() function or other functions to reduce the possibility of breaking the API. However, I understand this might not be the best way to do it, I am open to modify this if needed.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds a normalized TP center-offset RMSE metric to YOLO detection validation to quantify localization error beyond mAP. 📏

### 📊 Key Changes
-Computes per-true-positive center distance (normalized by image diagonal) for matches at IoU ≥ 0.5 in `ultralytics/models/yolo/detect/val.py`.
-Extends `DetMetrics` stats to store `tp_center_offset`, computes `center_rmse`, and exposes it via `keys()`, `mean_results()`, and `results_dict()` in `ultralytics/utils/metrics.py`.
-Adjusts `plot_results()` subplot layout logic to handle dynamic column counts more robustly in `ultralytics/utils/plotting.py`.

### 🎯 Purpose & Impact
-Provides an additional, interpretable localization-quality signal (normalized center RMSE) that complements mAP/IoU-based reporting.
-Helps compare models where mAP is similar but center placement accuracy differs, useful for downstream tracking/robotics-style use cases.
-Potential integration note: `results_dict()` currently appears to append `center_rmse` twice (once via `mean_results()` and again via `self.center_rmse`), which may misalign values vs. keys and should be double-checked before merge.

jmartinpizarro committed:
feat: implement rmse diagonal normalised metric

jmartinpizarro committed:
hotfix: plotting columns bug

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

UltralyticsAssistant reviewed (commented):
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Good addition overall: the metric is computed and threaded through stats/keys/means cleanly, and the plotting layout fix is directionally correct. The main blocker is in `DetMetrics.results_dict()` where `center_rmse` is appended twice, which will mis-report `fitness`. There are also a couple of robustness concerns around `tp_center_offset`’s type in `process()` and the TP matching implementation potentially not preserving IoU-priority matching.

💬 Posted 5 inline comments

jmartinpizarro committed:
hotfix: fix minor issues related with plotting

jmartinpizarro committed:
feat: changes in the tp calculation

Fix previous errors, such as precomputing the IoU again or assuming a hardcoded 0.5mAP for calculating (again) the TP-GT relationships.

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

jmartinpizarro committed:
hotfix: solve test problems with non-initialise values for the rest of classes

Computing twice this is not probably the best way to do it. However, I think it the best way of avoiding tests without modifying other major things that should not be touched

jmartinpizarro committed:
hotfix: error with tests

I should probably look for a refactor in process() method, but idk how. I'll wait for further comments or to tomorrow morning

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

jmartinpizarro committed:
feat: update the metrics calculation and extension

Now it is able to calculate TP-RMSE for detect and OBB models. Segmentations models are not included, as it wouldn't make any sense to use it. As the SegMetrics inherits from DetMetrics class, I needed to modify how the rmse is calculated and projected during inheritance in order to avoid non-useful metrics in other models.

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

jmartinpizarro committed:
hotfix: fix test

Basically, I was not checking for the shape of the pred_boxes. This could generate non OBB values tha, when reshaped, could not be resolved.

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

jmartinpizarro committed:
hotfix: avoid double execution of _process_batch()

glenn-jocher commented:
You’re reading it right: `_process_batch()` only returns a per-detection correctness matrix across `self.iouv` (e.g. 10 IoU thresholds), but it deliberately does **not** expose the GT↔prediction pairing, so you can’t recover “which GT matched this TP” from `tp` alone without computing/recording matches somewhere (see [`DetectionValidator._process_batch()`](https://docs.ultralytics.com/reference/models/yolo/detect/val/)).  

The lowest-risk approach is to keep `tp` unchanged and add an **extra** returned field (or compute `tp_center_offset` right before `update_stats`) by reusing the same IoU you already have and mirroring the “unique match” selection used in [`BaseValidator.match_predictions()`](https://docs.ultralytics.com/reference/engine/validator/) or the matching pattern in [`ConfusionMatrix.process_batch()`](https://docs.ultralytics.com/reference/utils/metrics/#ultralytics.utils.metrics.ConfusionMatrix.process_batch); adding a new dict key here is internal wiring and shouldn’t be considered an API break as long as existing `tp` behavior stays identical.

```python
# inside DetectionValidator._process_batch (keep existing tp exactly as-is)
iou = box_iou(batch["bboxes"], preds["bboxes"])
iou = iou * (batch["cls"][:, None] == preds["cls"])  # zero wrong-class pairs

thr = float(self.iouv[0])  # don't hardcode 0.5
x = torch.where(iou >= thr)

match_gt = torch.full((preds["cls"].shape[0],), -1, device=preds["cls"].device, dtype=torch.long)
if x[0].numel():
    matches = torch.stack(x, 1)  # [K, 2] -> (gt_i, det_i)
    m_iou = iou[x[0], x[1]]
    order = m_iou.argsort(descending=True)
    matches = matches[order]
    # greedily keep unique dets then unique gts (same idea as match_predictions)
    _, u_det = torch.unique(matches[:, 1], return_inverse=False, return_counts=False, sorted=False, return_index=True)
    matches = matches[u_det]
    _, u_gt = torch.unique(matches[:, 0], return_inverse=False, return_counts=False, sorted=False, return_index=True)
    matches = matches[u_gt]
    match_gt[matches[:, 1]] = matches[:, 0]
```

With `match_gt` you can compute a per-detection center offset only where `match_gt >= 0`, and pass that array through stats (always present, even for empty batches) via [`DetMetrics.update_stats()`](https://docs.ultralytics.com/reference/utils/metrics/#ultralytics.utils.metrics.DetMetrics.update_stats).

jmartinpizarro committed:
feat: fix test and change GT-preds calculation

Some simple tests for checking that the metrics do not crash. Furthermore, I have changed the way of computing the GT-Preds relationships (previously it was wrong). Now, I have added an internal wiring, computing the matches in _process_batch() and up-streaming them into the update_metrics function

jmartinpizarro committed:
hotfix: minor error in compatibility

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

jmartinpizarro committed:
hotfix: test related with dimensions

jmartinpizarro committed:
hotfix: i was not filtering

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

jmartinpizarro committed:
feat: obb did not have the calculation for match_gt

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
hotfix: improve test-coverage

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

Signed-off-by: Javier Martín Pizarro <117672712+jmartinpizarro@users.noreply.github.com>

jmartinpizarro committed:
hotfix: inheritance based bug propagate an attribute misvalued

jmartinpizarro committed:
hotfix: forgot to add task to other metrics

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

Signed-off-by: Javier Martín Pizarro <117672712+jmartinpizarro@users.noreply.github.com>

jmartinpizarro committed:
hotfix: remove the hardcoded flag for rmse metric added and calculation

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro commented:
@glenn-jocher Just a question, why was it putted to Stale?

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

Signed-off-by: Javier Martín Pizarro <117672712+jmartinpizarro@users.noreply.github.com>

jmartinpizarro committed:
hotfix: modify with respect to the previous merge commit, tp-rmse was not being appended

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

UltralyticsAssistant committed:
Merge branch 'main' into feature/tp-matched-rmse

glenn-jocher committed:
Merge remote-tracking branch 'origin/main' into feature/tp-matched-rmse

# Conflicts:
#	ultralytics/utils/metrics.py

glenn-jocher commented:
Hi @jmartinpizarro, I've merged the latest `main` into your branch and resolved the conflict in `ultralytics/utils/metrics.py`'s `update_image_metrics()`. `main` recently added an empty-GT perfect-score policy (P/R/F1 = 1.0 when an image has neither GT nor predictions), so I kept that alongside your `center_rmse` computation. Your branch is current again (0 commits behind `main`), and both `detect` and `obb` validation run cleanly end-to-end. The `keys()` / `mean_results()` / `results_dict()` outputs stay aligned and `center_rmse` is no longer appended twice, and your new tests pass.

One correctness issue to fix before merge, in `ultralytics/models/yolo/detect/val.py` (lines ~207-219): the center computation assumes boxes are either xyxy (4 cols) or a 9-col polygon, but for OBB `pbatch["bboxes"]`/`predn["bboxes"]` are **xywhr** (5 cols). They fall into the `else` branch `(boxes[:, :2] + boxes[:, 2:4]) / 2`, which for xywhr averages the center with `(w, h)` instead of taking the center directly, so two perfectly co-centered boxes of different sizes report a nonzero offset. For the 5-col xywhr case the center is simply `boxes[:, :2]`. (The `shape[1] == 9` branch is never reached in this pipeline.) Detect is unaffected and correct; only the OBB metric value is wrong.

Once that OBB center fix lands, this should be in good shape for review. Thanks for your patience on this one!

UltralyticsAssistant committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
hotfix: minor error in metrics related with obb calculation format

jmartinpizarro committed:
hotfix: no need of doing the mean

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro commented:
That should be it!

glenn-jocher commented:
Thanks, I see the update. CI is still running, so let’s wait for it to finish; if it stays green and the OBB `xywhr` center path is covered by the tests, this looks ready for maintainer review from my side.

jmartinpizarro committed:
feat: refactor and tests for the new loic metric

jmartinpizarro commented:
Added the tests. They work fine - in my local. If the CI/CD works fine I think everything you asked for has been done.

jmartinpizarro committed:
hotifx: docstring

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

glenn-jocher commented:
Thanks @jmartinpizarro, CI is green now, so the requested OBB `xywhr` fix and tests look complete from my side. This is ready for maintainer review; please just keep the branch synced with `main` if anything changes before review.

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

UltralyticsAssistant committed:
Merge branch 'main' into feature/tp-matched-rmse

UltralyticsAssistant committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

UltralyticsAssistant committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

glenn-jocher committed:
Merge main into feature/tp-matched-rmse

UltralyticsAssistant committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro committed:
Merge branch 'main' into feature/tp-matched-rmse

jmartinpizarro commented:
Roger! Thanks for all the help given!

glenn-jocher commented:
Hi @jmartinpizarro, thank you for the contribution and the thorough work on the metric! After review we've decided not to add TP-RMSE to the core validation metrics — the existing mAP/precision/recall suite covers our supported evaluation protocols, and each additional metric carries permanent maintenance and API surface. Closing as not planned. For specialized localization-error analysis, computing it externally from `results` / prediction JSON is fully supported. We appreciate the effort and hope to see more contributions from you!
